packages/next/client/components/reducer.ts

```diff
@@ -494,6 +499,42 @@ function shouldHardNavigate(
   )
 }
 
+function isNavigatingToNewRootLayout(
+  currentTree: FlightRouterState,
+  nextTree: FlightRouterState
+): boolean {
+  // Compare segments
+  const currentTreeSegment = currentTree[0]
+  const nextTreeSegment = nextTree[0]
+  if (Array.isArray(currentTreeSegment) && Array.isArray(nextTreeSegment)) {
+    // Compare dynamic param name and type but ignore the value
+    if (
+      currentTreeSegment[0] !== nextTreeSegment[0] ||
+      currentTreeSegment[2] !== nextTreeSegment[2]
+    ) {
+      return true
+    }
+  } else if (currentTreeSegment !== nextTreeSegment) {
+    return true
+  }
+
+  // Current tree root layout found
+  if (currentTree[4]) {
+    // Is navigating to new root layout if next tree doesn't have its root layout in the same segment
+    return !nextTree[4]
+  }
+  // Next tree has root layout where current tree doesn't
+  if (nextTree[4]) {
+    return true
+  }
+  // We can't assume it's `parallelRoutes.children` here in case the root layout is `app/@something/layout.js`
+  // But it's not possible to be more than one parallelRoutes before the root layout is found
+  const currentTreeChild = Object.values(currentTree[1])[0]
```

Comment:
Not needed for this PR but needed when we ship parallel routes so good to leave a todo
```suggestion
  // TODO-APP: change to traverse all parallel routes
  const currentTreeChild = Object.values(currentTree[1])[0]
```

Reply:
It shouldn't be possible to be several parallel routes before the root layout? 🤔